Get PX4 compile on ARM GCC 9

MaEtUgR opened the pull request:
**Describe problem solved by this pull request**
Follow up to https://github.com/PX4/Firmware/pull/13375 and https://github.com/PX4/Firmware/pull/13448.

**Describe your solution**
There are less problems e.g. the new NuttX version is compatible. As discussed with @dagar the main remaining blocking issue is the linker script / flash region error like last time. It seems out of my reach to solve this by myself.

- [x] All binary output including io `overflows end of region flash` for every single section. This is probably due to `warning: creating a segment to contain the file and program headers outside of any MEMORY region`. Maybe the additional "default" section has to be disabled or placed in the linker file (thanks for input by @dinomani).
```
/usr/lib/gcc/arm-none-eabi/9.2.0/../../../../arm-none-eabi/bin/ld.gold: warning: creating a segment to contain the file and program headers outside of any MEMORY region
/usr/lib/gcc/arm-none-eabi/9.2.0/../../../../arm-none-eabi/bin/ld.gold: error: section .text overflows end of region flash
/usr/lib/gcc/arm-none-eabi/9.2.0/../../../../arm-none-eabi/bin/ld.gold: error: section .init_section overflows end of region flash
/usr/lib/gcc/arm-none-eabi/9.2.0/../../../../arm-none-eabi/bin/ld.gold: error: address 0x82e9bd4 is not within region flash
/usr/lib/gcc/arm-none-eabi/9.2.0/../../../../arm-none-eabi/bin/ld.gold: error: section __param overflows end of region flash
/usr/lib/gcc/arm-none-eabi/9.2.0/../../../../arm-none-eabi/bin/ld.gold: error: section .ARM.extab overflows end of region flash
/usr/lib/gcc/arm-none-eabi/9.2.0/../../../../arm-none-eabi/bin/ld.gold: error: section .ARM.exidx overflows end of region flash
/usr/lib/gcc/arm-none-eabi/9.2.0/../../../../arm-none-eabi/bin/ld.gold: error: section .data overflows end of region flash
```

@dagar I had to comment the `--print-memory-usage` since it seems the parameter name is now different. We need to add this info back, it's very useful.

**Test data / coverage**
Untested because of linker error.

**Additional context**
Trying CI build on Windows now using the new toolchain version 0.9 with ARM GCC 9 from https://github.com/PX4/windows-toolchain/releases/tag/v0.9

julianoes commented:
Works for me for Pixracer:

```
WARN  [shutdown] Reboot NOW.
[boot] Fault Log info File No 4 Length 3177 flags:0x01 state:1
[boot] Fault Log is Armed
sercon: Registering CDC/ACM serial driver
sercon: Successfully registered the CDC/ACM serial driver
HW arch: PX4_FMU_V4
FW git-hash: f5dbc184f6db04891d27d4991a40ff8f7c9e3747
FW version: 1.11.0 80 (17498240)
FW git-branch: pr-arm-gcc-9-new
OS: NuttX
OS version: Release 8.2.0 (134349055)
OS git-hash: fdf1837077104e80912a2c46ff159fdacc8b06f9
Build datetime: Mar 31 2020 11:17:55
Build uri: localhost
Toolchain: GNU GCC, 9.2.1 20191025 (release) [ARM/arm-9-branch revision 277599]
PX4GUID: 000100000000303833363335510d00300041
MCU: STM32F42x, rev. 3
nsh: mount: mount failed: No such device
nsh: mkfatfs: mkfatfs failed: No such device
INFO  [param] selected parameter default file /fs/mtd_params
INFO  [tune_control] Publishing standard tune 16
Board defaults: /etc/init.d/rc.board_defaults
WARN  [dataman] Could not open data manager file /fs/microsd/dataman
ERROR [dataman] dataman start failed
WARN  [rgbled] no RGB led on bus #1
WARN  [rgbled_ncp5623c] no RGB led on bus #1
nsh: rgbled_pwm: command not found
Board sensors: /etc/init.d/rc.board_sensors
reset done, 50 ms
WARN  [hmc5883] no device on bus 1 (type: 2)
ERROR [hmc5883] bus 2 not started
ERROR [hmc5883] bus 2 not started
WARN  [lis3mdl] no device on bus 2
INFO  [ist8310] no device on bus 1
WARN  [qmc5883] no device on bus 1 (type: 2)
WARN  [rm3100] no device on bus 2
MS5611_SPI on SPI bus 2 at 3 (20000 KHz)
HMC5883_SPI on SPI bus 1 at 5 (11000 KHz)
WARN  [hmc5883] no device on bus 1 (type: 3)
ERROR [hmc5883] bus 3 not started
ERROR [hmc5883] bus 3 not started
WARN  [mpu6000] no device on bus #3 (SPI1)
WARN  [mpu6000] no device on bus #3 (SPI1)
MPU9250 on SPI bus 1 at 4 (1000 KHz)
WARN  [mpu9250] AK8963: bad id 163 retries 9
WARN  [mpu9250] AK8963: bad id 163 retries 8
WARN  [mpu9250] AK8963: bad id 163 retries 7
WARN  [mpu9250] AK8963: bad id 163 retries 6
WARN  [mpu9250] AK8963: bad id 163 retries 5
WARN  [mpu9250] AK8963: bad id 163 retries 4
WARN  [mpu9250] AK8963: bad id 163 retries 3
WARN  [mpu9250] AK8963: bad id 163 retries 2
Board extras: /etc/init.d/rc.board_mavlink
INFO  [mavlink] mode: Config, data rate: 800000 B/s on /dev/ttyACM0 @ 57600B
INFO  [mavlink] mode: Normal, data rate: 20000 B/s on /dev/ttyS0 @ 921600B
WARN  [mavlink] hardware flow control not supported
ERROR [mavlink] DM_KEY_MISSION_STATE lock failed
ERROR [mavlink] offboard mission init failed (-1)
Starting Main GPS on /dev/ttyS3
Starting MAVLink on /dev/ttyS1
nsh: mavlink: too many arguments
INFO  [mavlink] mode: Normal, data rate: 1200 B/s on /dev/ttyS1 @ 57600B
INFO  [init] Mixer: /etc/mixers/quad_x.main.mix on /dev/pwm_output0
Board extras: /etc/init.d/rc.board_extras
INFO  [logger] logger started (mode=all)
INFO  [logger] log root dir created: /fs/microsd/log

NuttShell (NSH)
nsh> INFO  [ecl/EKF] reset position to last known position
INFO  [ecl/EKF] reset velocity to zero
INFO  [ecl/EKF] 6108718: EKF aligned, (baro height, IMU buf: 22, OBS buf: 14)
INFO  [frsky_telemetry] Scanning timeout: exiting
```

TSC21 commented:
@MaEtUgR can you rebase this and see where it goes? Note that you need to point to the container tag that still has the gcc 9 toolchain.

julianoes commented:
And seems to work on Pixhawk 4:
```
INFO  [ist8310] no device on bus 1
INFO  [ist8310] no device on bus 2
WARN  [hmc5883] no device on bus 1 (type: 2)
WARN  [hmc5883] no device on bus 2 (type: 2)
WARN  [hmc5883] no device on bus 4 (type: 2)
ERROR [hmc5883] bus 2 not started
ERROR [hmc5883] bus 2 not started
WARN  [qmc5883] no device on bus 1 (type: 2)
WARN  [qmc5883] no device on bus 2 (type: 2)
WARN  [qmc5883] no device on bus 4 (type: 2)
WARN  [lis3mdl] no device on bus 2
WARN  [lis3mdl] no device on bus 2
WARN  [lis3mdl] no device on bus 2
nsh: icm20948: command not found
WARN  [ak09916] AK09916 mag init failed
ERROR [ak09916] driver start failed
IST8310 on I2C bus 3 at 0x0e (bus: 100 KHz, max: 400 KHz)
MS5611_SPI on SPI bus 4 at 0 (20000 KHz)
WARN  [rm3100] no device on bus 2
WARN  [rm3100] no device on bus 2
WARN  [rm3100] no device on bus 2
WARN  [rm3100] no device on bus 1
Board extras: /etc/init.d/rc.board_mavlink
INFO  [mavlink] mode: Config, data rate: 800000 B/s on /dev/ttyACM0 @ 57600B
Starting Main GPS on /dev/ttyS0
Starting MAVLink on /dev/ttyS1
nsh: mavlink: too many arguments
INFO  [mavlink] mode: Normal, data rate: 1200 B/s on /dev/ttyS1 @ 57600B
INFO  [px4io] default PWM output device
INFO  [init] Mixer: /etc/mixers/quad_x.main.mix on /dev/pwm_output0
INFO  [init] Mixer: /etc/mixers/pass.aux.mix on /dev/pwm_output1
INFO  [logger] logger started (mode=all)

NuttShell (NSH)
nsh> INFO  [ecl/EKF] reset position to last known position
INFO  [ecl/EKF] reset velocity to zero

nsh>
nsh>
nsh>
nsh> px4INFO  [ecl/EKF] 14685700: EKF aligned, (baro height, IMU buf: 22, OBS buf: 14)
io status
WARN  [px4io] loaded
protocol 4 hardware 2 bootloader 3 buffer 64B crc 0xa2823dd9
8 controls 8 actuators 18 R/C inputs 2 analog inputs 0 relays
216 bytes free
status 0x2741 OUTPUTS_ARMED SAFETY_SAFE RC_FAIL FMU_OK MIXER_OK ARM_SYNC INIT_OK
alarms 0x0030 FMU_LOST RC_LOST
vservo 2 mV vservo scale 10000
vrssi 1164
actuators -10000 -10000 -10000 -10000 0 0 0 0
servos 900 900 900 900 0 0 0 0
reversed outputs: [________] trims: r:   0.0000 p:   0.0000 y:   0.0000
0 raw R/C inputs
R/C flags: 0x0000
mapped R/C inputs 0x0000
ADC inputs 1 1099
features 0x0008 RSSI_ADC
arming 0x0861 FMU_DISARMED FMU_NOT_PREARMED IO_ARM_OK INAIR_RESTART_OK ALWAYS_PWM_ENABLE
rates 0x000f default 50 alt 400 sbus 72
debuglevel 0
controls 0: 249 -1552 -20 0 0 0 0 -10000
controls 1: 0 0 0 0 0 0 0 0
controls 2: 0 0 0 0 0 0 0 0
controls 3: 0 0 0 0 0 0 0 0
input 0 min 1000 center 1500 max 2000 deadzone 10 assigned 255 options 0x0000
input 1 min 1000 center 1500 max 2000 deadzone 10 assigned 255 options 0x0000
input 2 min 1000 center 1500 max 2000 deadzone 10 assigned 255 options 0x0000
input 3 min 1000 center 1500 max 2000 deadzone 10 assigned 255 options 0x0000
input 4 min 1000 center 1500 max 2000 deadzone 10 assigned 255 options 0x0000
input 5 min 1000 center 1500 max 2000 deadzone 10 assigned 255 options 0x0000
input 6 min 1000 center 1500 max 2000 deadzone 10 assigned 255 options 0x0000
input 7 min 1000 center 1500 max 2000 deadzone 10 assigned 255 options 0x0000
input 8 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
input 9 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
input 10 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
input 11 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
input 12 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
input 13 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
input 14 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
input 15 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
input 16 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
input 17 min 1000 center 1500 max 2000 deadzone 0 assigned 255 options 0x0000
failsafe 900 900 900 900 1500 1500 0 0
disarmed values 900 900 900 900 0 0 0 0
```

MaEtUgR commented:
I can confirm I had it also working on pixhawk4 yesterday built using a manually adjusted Windows Cygwin toolchain with `gcc-arm-none-eabi-9-2019-q4-major-win32`. I don't know what changed since my last tests 2 months ago at least I didn't change the branch.

MaEtUgR commented:
I rebased and did some tests again:
- With `wget https://armkeil.blob.core.windows.net/developer/Files/downloads/gnu-rm/9-2019q4/gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2` it all works fine. With the packaged compilers not. They produce the overflow errors. I reverted the `arch.sh` change FYI @julianoes. We can switch to the latest version in the setup script later when we have more testing.
- The nice `--print-memory-usage` works as well now and I put it back in place.

CI is missing atm but should now pass also with version 9. Reviews and comments welcome.

TSC21 commented:
> I rebased and did some tests again:
> 
> * With `wget https://armkeil.blob.core.windows.net/developer/Files/downloads/gnu-rm/9-2019q4/gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2` it all works fine. With the packaged compilers not. They produce the overflow errors. I reverted the `arch.sh` change FYI @julianoes. We can switch to the latest version in the setup script later when we have more testing.
> * The nice `--print-memory-usage` works as well now and I put it back in place.
> 
> CI is missing atm but should now pass also with version 9. Reviews and comments welcome.

So let's update the containers :+1:

MaEtUgR commented:
@dagar GCC 9 vs 7 seems to
- save 6288 Bytes of flash on current master `fmu-v5` but use 3968 Bytes more SRAM
- save 4056 Bytes of flash on current master `fmu-v2` and use the same amount of SRAM

Is the increase in RAM usage on fmu-v5 just a more accurate prediction?

fmu-**v5**, **7**-2017-q4-major:
```
[222/224] Linking CXX executable px4_io-v2_default.elf
Memory region         Used Size  Region Size  %age Used
           flash:       59064 B        60 KB     96.13%
            sram:        4036 B         8 KB     49.27%
[224/224] Creating /home/maetugr/Firmware/build/px4_fmu-v5_default/external/Build/px4io_firmware/px4_io-v2_default.px4
[1349/1351] Linking CXX executable px4_fmu-v5_default.elf
Memory region         Used Size  Region Size  %age Used
            itcm:          0 GB      2016 KB      0.00%
           flash:     1763633 B      2016 KB     85.43%
            dtcm:          0 GB       128 KB      0.00%
           sram1:       46136 B       368 KB     12.24%
           sram2:          0 GB        16 KB      0.00%
```

fmu-**v5**, **9**-2019-q4-major:
```
[222/224] Linking CXX executable px4_io-v2_default.elf
Memory region         Used Size  Region Size  %age Used
           flash:       59508 B        60 KB     96.86%
            sram:        4036 B         8 KB     49.27%
[224/224] Creating /home/maetugr/Firmware/build/px4_fmu-v5_default/external/Build/px4io_firmware/px4_io-v2_default.px4
[1349/1351] Linking CXX executable px4_fmu-v5_default.elf
Memory region         Used Size  Region Size  %age Used
            itcm:          0 GB      2016 KB      0.00%
           flash:     1757345 B      2016 KB     85.13%
            dtcm:          0 GB       128 KB      0.00%
           sram1:       50104 B       368 KB     13.30%
           sram2:          0 GB        16 KB      0.00%
```

fmu-**v2** before **7**-2017-q4-major:
```
[222/224] Linking CXX executable px4_io-v2_default.elf
Memory region         Used Size  Region Size  %age Used
           flash:       59064 B        60 KB     96.13%
            sram:        4036 B         8 KB     49.27%
[224/224] Creating /home/maetugr/Firmware/build/px4_fmu-v2_default/external/Build/px4io_firmware/px4_io-v2_default.px4
[734/736] Linking CXX executable px4_fmu-v2_default.elf
Memory region         Used Size  Region Size  %age Used
           flash:     1022781 B      1008 KB     99.09%
            sram:       26592 B       192 KB     13.53%
          ccsram:          0 GB        64 KB      0.00%
```
fmu-**v2** after **9**-2019-q4-major:
```
[222/224] Linking CXX executable px4_io-v2_default.elf
Memory region         Used Size  Region Size  %age Used
           flash:       59508 B        60 KB     96.86%
            sram:        4036 B         8 KB     49.27%
[224/224] Creating /home/maetugr/Firmware/build/px4_fmu-v2_default/external/Build/px4io_firmware/px4_io-v2_default.px4
[734/736] Linking CXX executable px4_fmu-v2_default.elf
Memory region         Used Size  Region Size  %age Used
           flash:     1018725 B      1008 KB     98.70%
            sram:       26592 B       192 KB     13.53%
          ccsram:          0 GB        64 KB      0.00%
```

shrit commented:
@MaEtUgR @julianoes I am getting exactly the same error as the first post on my Arch Linux with arm-none-eabi-ld.gold (2.34) and (2.32) when trying to compile PX4 for my crazyflie 2.1

TSC21 commented:
> @MaEtUgR @julianoes I am getting exactly the same error as the first post on my Arch Linux with arm-none-eabi-ld.gold (2.34) and (2.32) when trying to compile PX4 for my crazyflie 2.1
> In addition, I am using the master version of the Firmware

Are you compiling the Firmware already with this change in?

shrit commented:
@TSC21 Yes, I have just downloaded the firmware from github and I am using the master version

TSC21 commented:
We are still validiting this PR and as soon as it is validated the documentation and setup scripts will be updated. Did you use the install script we provide for arch? We use the same toolset as for Ubuntu for example (https://github.com/PX4/Firmware/blob/master/Tools/setup/arch.sh#L93). You might wanna check that first.

shrit commented:
@TSC21 I was able to compile using gcc 2017 from the `arch.sh` script. I understand that this pull request needs to be tested but since it has been merged I thought it has resolved the problem for all the arm devices.

TSC21 commented:
> @TSC21 I was able to compile using gcc 2017 from the `arch.sh` script. I understand that this pull request needs to be tested but since it has been merged I thought it has resolved the problem for all the arm devices.

When I say tested, I mean flight tested (https://github.com/PX4/Firmware/issues/14580). This PR is merged because it builds correctly with `gcc-arm-none-eabi-9-2019-q4-major`, which is what you are supposed to be installing on you Arch system if you want to build against GCC 9.

shrit commented:
Ok, So Why it did not work on my Arch with `gcc-arm-none-eabi-9-2019-q4-major` with `ld 2.34` and `ld 2.32`??
I had to remove the latest cross-compile toolchain and test it with gcc2017 in order to work

TSC21 commented:
> Ok, So Why it did not work on my Arch with `gcc-arm-none-eabi-9-2019-q4-major` with `ld 2.34` and `ld 2.32`??
> I had to remove the latest cross-compile toolchain and test it with gcc2017 in order to work

Sorry but that I cannot answer. I don't know enough about ld.

MaEtUgR commented:
@shrit Sorry I didn't see this earlier. I had the same problem. If you download the compiler from here https://launchpad.net/gcc-arm-embedded it contains everything you need and ARM GCC 9 on arch works. If you install it from the package repository that golden linker is for some reason missing. There seems to be a lot of people in forums/stackoverflow having this problem.

shrit commented:
@MaEtUgR Thanks for your reply.
No problem, I have used the 2017 version of gcc and the linker. I was compiling the firmware for Crazyflie 2.1 but I realized later that the firmware do not support this board.

MaEtUgR commented:
@shrit ok, the differences seem marginal: https://www.bitcraze.io/2019/02/the-crazyflie-2-1-is-here/ I bet you get it working if you're up to a bit of development. It's just different sensors that PX4 has drivers for so you'll probably just have to start the right ones.

shrit commented:
@MaEtUgR Thanks for your reply.
Exactly, I have noticed that the difference is only related to sensors.
Do you have a starting point? I have noticed that the boards are supported in `board` directory, but I am not sure that this is the only thing to modify some `#defines` related to senors.

MaEtUgR commented:
@shrit Here's where the board specific drivers are started: https://github.com/PX4/Firmware/blob/master/boards/bitcraze/crazyflie/init/rc.board_sensors#L8-L12
I don't know what else is necessary. Probably setting the right interface pins since I hope they didn't stick to I2C for the IMU but switched to SPI.
